### ISSUE TITLE
Servo does not break Thai text line properly.

### ISSUE THREAD

**Issue by veer66** [servo/servo#13088]
I use Servo a0f45c6850b6 on Ubuntu 16.04. The text "การบ้านการบ้านการบ้าน" is supposed to be split but it isn't.

![servo_break1](https://cloud.githubusercontent.com/assets/370688/18032123/6e82e87a-6d2e-11e6-9b4f-eca71e9d2c9f.png)

But it is supposed to look like this:

![ffbreak1](https://cloud.githubusercontent.com/assets/370688/18032125/73c892da-6d2e-11e6-88bf-5e8e68dff51b.png)

The page for testing is here: http://file.veerkesto.net/servo_test/thai_line_break.html

**Comment by Manishearth** [comment]
cc @glennw

**Comment by mbrubeck** [comment]
Thai line breaking require language-specific dictionaries or text analysis. [UAX#14][1] says:

> The third style is used for scripts such as Thai, which do not use spaces, but which restrict word-breaks to syllable boundaries, the determination of which requires knowledge of the language comparable to that required by a hyphenation algorithm. Such an algorithm is beyond the scope of the Unicode Standard.

Thai line breaking is implemented by libraries like [libthai][2]. To support this in Servo, we would need either a Rust binding to an existing library, or a Rust implementation of the algorithm. Then we would need to call it from [`TextRun::break_and_shape`][3] when shaping text using the Thai script.

[1]: http://www.unicode.org/reports/tr14/tr14-17.html#BreakOpportunities
[2]: https://github.com/tlwg/libthai
[3]: https://github.com/servo/servo/blob/6a0c70a22269cbec6b72cfa8db499838bb00a0c3/components/gfx/text/text_run.rs#L196-L233

**Comment by veer66** [comment]
Is this approach from 2007 still preferred https://bugzilla.mozilla.org/show_bug.cgi?id=336959 ?

**Comment by mbrubeck** [comment]
Yes, that's similar to what I suggested in my previous comment. It looks like Gecko currently uses Pango on GTK platforms, Uniscribe on Windows, system APIs on macOS, and its own [rule-based line breaker](https://hg.mozilla.org/mozilla-central/file/29beaebdfaccbdaeb4c1ee5a43a9795ab015ef49/intl/lwbrk/rulebrk.c) on other platforms.

**Comment by veer66** [comment]
Since obtaining offsetWidth of span does not work yet #12939, I suppose that word break tests do not work as expected.

**Comment by khaledhosny** [comment]
I don’t think Gecko uses Pango or Uniscribe for anything right now. ICU provides line breaking support that covers Thai and other languages, and I think Gecko is moving to replace its line breaker with ICU’s.

**Comment by nox** [comment]
Sample is gone.

**Comment by veer66** [comment]
I put a new url.

**Comment by mrobinson** [comment]
This seems to be working properly now:

```html
<!DOCTYPE html>
<div style="width: 20px; border: solid black 1px;">การบ้านการบ้านการบ้าน</div>
```

<img width="334" height="373" alt="Image" src="https://github.com/user-attachments/assets/7cff1232-0b92-4af5-971e-19f0e33c11e8" />
